pkg/handler/collector/oci/oci.go

```diff
@@ -0,0 +1,267 @@
+       ociRepoCollector.checkedDigest = o.checkedDigest[repo]
+       ociRepoCollector.RetrieveArtifacts(ctx, docChannel)
+       o.checkedDigest[repo] = ociRepoCollector.checkedDigest
+     }
+   }
+   return nil
+*/
+func NewOCICollector(ctx context.Context, repo string, tag string, poll bool, interval time.Duration) *ociCollector {
+	return &ociCollector{
+		repo:          repo,
+		tag:           tag,
+		checkedDigest: []string{},
+		poll:          poll,
+		interval:      interval,
+	}
+}
+
+// RetrieveArtifacts get the artifacts from the collector source based on polling or one time
+func (o *ociCollector) RetrieveArtifacts(ctx context.Context, docChannel chan<- *processor.Document) error {
+	if o.poll {
+		if o.tag != "" {
+			return errors.New("image tag should not specified when using polling")
+		}
+		for {
+			err := o.getTagsAndFetch(ctx, docChannel)
+			if err != nil {
+				return err
+			}
+			// set interval to about 5 mins or more
+			time.Sleep(o.interval)
+		}
+	} else {
+		if o.tag == "" {
+			return errors.New("image tag not specified to fetch")
+		}
+		err := o.getTagsAndFetch(ctx, docChannel)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (o *ociCollector) getTagsAndFetch(ctx context.Context, docChannel chan<- *processor.Document) error {
+	rcOpts := []regclient.Opt{}
+	rcOpts = append(rcOpts, regclient.WithDockerCreds())
+	rcOpts = append(rcOpts, regclient.WithDockerCerts())
+
+	if o.tag != "" {
+		imageTag := fmt.Sprintf("%v:%v", o.repo, o.tag)
+		r, err := ref.New(imageTag)
+		if err != nil {
+			return err
+		}
+
+		rc := regclient.New(rcOpts...)
+		defer rc.Close(ctx, r)
+
+		err = o.fetchOCIArtifacts(ctx, rc, r, docChannel)
+		if err != nil {
+			return err
+		}
+	} else {
+		r, err := ref.New(o.repo)
+		if err != nil {
+			return err
+		}
+
+		rc := regclient.New(rcOpts...)
+		defer rc.Close(ctx, r)
+
+		tags, err := rc.TagList(ctx, r)
+		if err != nil {
+			return fmt.Errorf("reading tags for %s: %w", o.repo, err)
+		}
+
+		for _, tag := range tags.Tags {
+			if !strings.HasSuffix(tag, "sbom") && !strings.HasSuffix(tag, "att") && !strings.HasSuffix(tag, "sig") {
+				imageTag := fmt.Sprintf("%v:%v", o.repo, tag)
+				r, err := ref.New(imageTag)
+				if err != nil {
+					return err
+				}
+				err = o.fetchOCIArtifacts(ctx, rc, r, docChannel)
+				if err != nil {
+					return err
+				}
+			}
+		}
+	}
+	return nil
+}
+
+func (o *ociCollector) fetchOCIArtifacts(ctx context.Context, rc *regclient.RegClient, image ref.Ref, docChannel chan<- *processor.Document) error {
+	logger := logging.FromContext(ctx)
+
+	// attempt to request only the headers, avoids Docker Hub rate limits
+	m, err := rc.ManifestHead(ctx, image)
+	if err != nil {
+		return err
+	}
+
+	if m.IsList() {
+		m, err := rc.ManifestGet(ctx, image)
+		if err != nil {
+			return err
+		}
+		pl, _ := manifest.GetPlatformList(m)
+		for _, p := range pl {
+			desc, err := manifest.GetPlatformDesc(m, p)
```

Comment:
this may be a silly question, but when a `.sbom` or `.att` is uploaded by cosign for a multi arch/platform image, i assume that because it does that based on the digest of the container image that it has one for each arch/platform image? is that the case?

Reply:
Yup that is correct. See here as an example I am using for the unit test. https://hub.docker.com/r/ppatel1989/go-multi-test/tags